### PR TITLE
`bitmap-copy`: upgrade into Bitmap images copying/pasting

### LINKED ISSUE #3788: `bitmap-copy`: allow pasting images from the global clipboard
### Is your request related to a problem? Please describe.

I would like to be able to paste images from other software directly into the Scratch paint editor, but I can't without saving the image first.

### Describe the solution you'd like

Allow `bitmap-copy` to paste images from the global clipboard into the bitmap paint editor.

### Describe alternatives you've considered

Making it a new add-on or saving the images I want to import to the downloads folder, importing it into Scratch and then deleting them.

### Additional context

Are writing and reading the clipboard two separate permissions or are they combined?

### PR DESCRIPTION
**Pull request by CST1229**
Resolves #3788

### Changes

Adds a feature to `bitmap-copy` that allows users to paste in images, either with an option in the creation menu or by pressing Ctrl+V after clicking on the list of costumes, sprite pane or stage pane.
Also fixes a bug in `better-img-uploads` that prevented it from working on the backdrop creation menu.
Technical: Adds a `spriteCreationMenu` space, used by `better-img-uploads` and `bitmap-copy`.

### Reason for changes

Sometimes you just don't want to download a file to upload it.

### Tests

Only partially tested, seems to work. Returning to project page and back prevents the costume pane buttons from both this addon and `better-img-uploads` from working, but I don't know how to fix it and it probably existed before this PR. Dynamic enable/disable seems to work.
https://user-images.githubusercontent.com/68464103/149924209-bd1460d4-d69b-43f9-8e3a-c8dc87860578.mp4
The addon _might_ need the Beta tag.
Also some of this code is probably bad as I didn't ensure super good code quality, and there are parts pasted from better-img-uploads that create the menu button.

### CODE REVIEW

**Comment by Samq64** [comment]
The addon name seems a bit weird. Maybe something like `Bitmap image copying and pasting` would be better?

**Comment by RedGuy12** [comment]
@samq64 we want to avoid verbs, remember


________________________________

**Comment by mxmou** [comment]
> @Samq64 we want to avoid verbs, remember

So "images copying" is a noun but "image copying" is a verb"?

**Comment by WorldLanguages** [comment]
Found a weird bug?

1. Go to https://scratch.mit.edu/projects/104/editor/ girl sprite, costumes tab
2. Try pasting with Ctrl+V, won't work
3. Create a new empty costume
4. Ctrl+V now works

**Comment by WorldLanguages** [comment]
By the way, I'm surprised we took this long to develop something like this - I was pretty sure you needed a server to paste images. I must have mistaken pasting url vs pasting an image.

**Comment by CST1229** [comment]
> Found a weird bug?
> 
>     1. Go to [scratch.mit.edu/projects/104/editor](https://scratch.mit.edu/projects/104/editor/) girl sprite, costumes tab
> 
>     2. Try pasting with Ctrl+V, won't work
> 
>     3. Create a new empty costume
> 
>     4. Ctrl+V now works

That's more of a UX issue - to prevent accidentally creating a costume/sprite, passive pasting only works if the last thing you clicked on was the costumes list (including create costume button), sprite pane (including create sprite button) or the stage pane (including create backdrop button). Since the Costumes tab button isn't one of those, it doesn't work if you clicked on it last (which is the case when going to the Costumes tab).
I'll add that in.

**Comment by CST1229** [comment]
The fix I just pushed still allows the costume editor to paste the Scratch clipboard alongside creating a costume, but I don't know how to fix that. Calling `e.preventDefault()` and `e.stopPropagation()` doesn't fix it.

**Comment by CST1229** [comment]
Also, should I add myself to the addon credits?

**Comment by RedGuy12** [comment]
> Also, should I add myself to the addon credits?

yes

File: addons/bitmap-copy/addon.json [hunk 1]
```diff
@@ -12,11 +12,34 @@
...
     }
   ],
+  "credits": [
+    {
+      "name": "CST1229",
+      "note": "Pasting images",
+      "id": "pasting",
+      "link": "https://github.com/CST1229"
+    }
+  ]
```
Comment (apple502j) [inline addons/bitmap-copy/addon.json:34]: It crashes 🤦🏼 

```suggestion
  ],
```

**Comment by apple502j** [comment]
@CST1229 It seems like transparency gets lost during the paste process.

Wikipedia logo on Scratch space, but transparent parts are filled black

**Comment by apple502j** [comment]
Okay, transparency issue seems to be a design flaw of Windows from quick web search. Oh well.

File: addons/bitmap-copy/addon.json (hunk 1 above)
Comment (CST1229) [reply addons/bitmap-copy/addon.json]: Didn't test, oops

**Review by apple502j — APPROVED** [review]

**Comment by CST1229** [comment]
Note that there are issues with passive paste when something is in the Scratch clipboard, so I think I'll have to work on that sometime.

**Comment by WorldLanguages** [comment]
Any updates? Maybe we could have garbo take a look?

**Comment by Weredime** [comment]
> Any updates? Maybe we could have garbo take a look?

cc @GarboMuffin

**Comment by WorldLanguages** [comment]
> Note that there are issues with passive paste when something is in the Scratch clipboard (both clipboards pasting at the same time), so I think I'll have to work on that sometime.

@CST1229 Could you post a written summary of what issues are still pending?

**Comment by CST1229** [comment]
> > Note that there are issues with passive paste when something is in the Scratch clipboard (both clipboards pasting at the same time), so I think I'll have to work on that sometime.
> 
> @CST1229 Could you post a written summary of what issues are still pending?

after previous commits:

- vanilla pasting still takes effect in both manual and passive paste, idk how to override
- passive paste switches tools, idk how to consistently fully instantly update the image
- maybe we should consider removing passive paste altogether, because it probably needs to override vanilla pasting with ctrl+v

**Comment by GrahamSH-LLK** [comment]
Is this abandoned?

**Comment by CST1229** [comment]
yea

**Comment by MegaBaconator** [comment]
aw man really needed this

**Pull request closed without merging**